src/Futhark/Actions.hs

```diff
@@ -255,11 +329,46 @@ compileMulticoreAction fcfg mode outpath =
           liftIO $ writeFile hpath $ cPrependHeader header
           liftIO $ writeFile cpath $ cPrependHeader impl
         ToExecutable -> do
-          liftIO $ writeFile cpath $ cPrependHeader $ MulticoreC.asExecutable cprog
-          runCC cpath outpath ["-O", "-std=c99"] ["-lm", "-pthread"]
+          liftIO $ writeFile cpath $ MulticoreC.asExecutable cprog
+          runCC cpath outpath ["-O3", "-std=c99"] ["-lm", "-pthread"]
         ToServer -> do
-          liftIO $ writeFile cpath $ cPrependHeader $ MulticoreC.asServer cprog
-          runCC cpath outpath ["-O", "-std=c99"] ["-lm", "-pthread"]
+          liftIO $ writeFile cpath $ MulticoreC.asServer cprog
+          runCC cpath outpath ["-O3", "-std=c99"] ["-lm", "-pthread"]
+
+compileMulticoreToWASMAction :: FutharkConfig -> CompilerMode -> FilePath -> Action MCMem
+compileMulticoreToWASMAction fcfg mode outpath =
+  Action
+    { actionName = "Compile to sequential C",
+      actionDescription = "Compile to sequential C",
+      actionProcedure = helper
+    }
+  where
+    helper prog = do
+      (cprog, jsprog, exps) <- handleWarnings fcfg $ MulticoreWASM.compileProg prog
+
+      case mode of
+        ToExecutable -> do
+          liftIO $ writeFile cpath $ MulticoreC.asExecutable cprog
```

Comment:
It's going to be the same function in the end, but I think it's more appropriate to say `MulticoreWASM.asExecutable` here.  Alternatively, remove the exports from the WASM modules.